x-pack/plugins/observability_solution/dataset_quality/server/services/data_telemetry/data_telemetry_service.ts

```diff
@@ -0,0 +1,267 @@
+import type { TelemetryPluginStart } from '@kbn/telemetry-plugin/server';
+
+import {
+  BREATHE_DELAY_MEDIUM,
+  BREATHE_DELAY_SHORT,
+  NON_LOG_SIGNALS,
+  EXCLUDE_ELASTIC_LOGS,
+  MAX_STREAMS_TO_REPORT,
+  LOGS_DATASET_INDEX_PATTERNS,
+  LOGS_DATA_TELEMETRY_TASK_TYPE,
+  TELEMETRY_TASK_INTERVAL,
+  LOGS_DATA_TELEMETRY_TASK_ID,
+  TELEMETRY_TASK_TIMEOUT,
+} from './constants';
+import {
+  getAllIndices,
+  addMappingsToIndices,
+  addNamespace,
+  groupStatsByPatternName,
+  getIndexBasicStats,
+  indexStatsToTelemetryEvents,
+  getIndexFieldStats,
+} from './helpers';
+
+import { DataTelemetryEvent } from './types';
+
+export class DataTelemetryService {
+  private readonly logger: Logger;
+  private isStopped = false;
+
+  private telemetryStart?: TelemetryPluginStart;
+
+  // @ts-ignore: Unused variable
+  private analytics?: AnalyticsServiceSetup;
+
+  // @ts-ignore: Unused variable
+  private isInProgress = false;
+
+  private isOptedIn?: boolean = true; // Assume true until the first check
+  private esClient?: ElasticsearchClient;
+
+  private run$ = defer(() => from(this.shouldCollectTelemetry())).pipe(
+    takeWhile(() => !this.isStopped),
+    tap((isOptedIn) => {
+      if (!isOptedIn) {
+        this.logTelemetryNotOptedIn();
+        this.isInProgress = false;
+      } else {
+        this.isInProgress = true;
+      }
+    }),
+    filter((isOptedIn) => isOptedIn),
+    exhaustMap(() => this.collectAndSend()),
+    tap(() => (this.isInProgress = false))
+  );
+
+  constructor(logger: Logger) {
+    this.logger = logger;
+  }
+
+  public setup(analytics: AnalyticsServiceSetup, taskManager: TaskManagerSetupContract) {
+    this.analytics = analytics;
+    this.registerTask(taskManager);
+  }
+
+  public async start(
+    telemetryStart?: TelemetryPluginStart,
+    core?: CoreStart,
+    taskManager?: TaskManagerStartContract
+  ) {
+    this.telemetryStart = telemetryStart;
+    this.esClient = core?.elasticsearch.client.asInternalUser;
+
+    if (taskManager) {
+      const taskInstance = await this.scheduleTask(taskManager);
+      if (taskInstance) {
+        this.logger.debug(`Task ${taskInstance.id} scheduled.`);
+      }
+    }
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  private registerTask(taskManager: TaskManagerSetupContract) {
+    const service = this;
+    taskManager.registerTaskDefinitions({
+      [LOGS_DATA_TELEMETRY_TASK_TYPE]: {
+        title: 'Logs Data Telemetry',
+        description:
+          'This task collects data telemetry for logs data and sends it to the telemetry service.',
+        timeout: `${TELEMETRY_TASK_TIMEOUT}m`,
+        maxAttempts: 1, // Do not retry
+
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            // Perform the work of the task. The return value should fit the TaskResult interface.
+            async run() {
+              const { state } = taskInstance;
```

Comment:
If you're not using `state` at all, you don't need to get it from the `taskInstance` nor return it as the result of the executor function.